### PR TITLE
Initial Pallas Matmul support in torch/_inductor.

### PR DESCRIPTION
**Pull request by yarongmu-google**
cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

### CODE REVIEW

File: torch/_inductor/choices.py [hunk 1]
```diff
@@ -162,10 +162,14 @@ def get_ktc(
...
-        cs = heuristic.get_template_configs(
-            kernel_inputs,
-            op_name,
+        configs = list(
+            heuristic.get_template_configs(
+                kernel_inputs,
+                op_name,
+            )
         )
+        print(f"Configs from heuristic for {template_name}: {configs}")
```
Comment (jansel) [inline torch/_inductor/choices.py]: Remove prints throughout

File: torch/_inductor/codegen/pallas_kernels/block_mm_template.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+"""
+
+MAIN_FUNCTION_TEMPLATE = """
+def <KERNEL_NAME>_main({{main_fn_args}}, stream=None):
+    # Enable JAX x64 mode for float64/int64 support
+    jax.config.update('jax_enable_x64', True)
+
+    # Move torch tensors to CPU, convert to numpy, then put on TPU device
+    x_tpu_jax = jax.device_put({{x}}.cpu().numpy(), device=jax.devices('tpu')[0])
+    y_tpu_jax = jax.device_put({{y}}.cpu().numpy(), device=jax.devices('tpu')[0])
```
Comment (jansel) [inline torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: I'm getting worried about spreading this CPU-tensor hack to more places.  It will have abysmal performance, so should really only be used to unblock other work -- but templates are mostly boilerplate (and they are optional in inductor) and doesn't unblock anything else, so I would leave that until we have a better way to call pallas kernels.

For GPU, we need to switch to calling pallas via `pl.kernel` + `plgpu.as_torch_kernel` for GPU see [mgpu_torch_test.py](https://github.com/jax-ml/jax/blob/71f1a99af4807b785c1d7786dedbe12d7d5e8eac/tests/pallas/mgpu_torch_test.py).  We should create a TPU version of that same API if it doesn't already exist.

FYI another way to do matmuls would be to set this flag:
https://github.com/pytorch/pytorch/blob/33d4cf4fcb7f0cba6191b242dae53b48057e05b9/torch/_inductor/config.py#L1485-L1501

Or you could leave them as a fallback op for now.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/choices.py (hunk 1 above)
Comment (yarongmu-google) [reply torch/_inductor/choices.py]: Done.

File: torch/_inductor/codegen/pallas_kernels/block_mm_template.py (hunk 2 above)
Comment (yarongmu-google) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: Agree. How about this? given that this is a sizable change and mainly aims to plubming Pallas through the non-element-wise path, let's merge this in. Next I can install torch_tpu (or toech_xla is torch_tpu really doesn't work) so we no longer have the CPU hacks?

Isolating env/dependency changes and code changes into 2 separate PRs make the PRs each much cleaner and with a single goal. 

This plumbing could be tested way more thoroughly etc etc and we will add these individually in different PRs - enabling clean encapsulation and easy review?

Comment (jansel) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: Is there a rush to land this one?  Does anything depend on it?

Once we get the proper plumbing this PR will need to change anyway.

Comment (yarongmu-google) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: Our goal was to get accuracy first, and then performance. This PR establishes that accuracy baseline, even though we used JAX to send the tensors to the TPU instead of torch_tpu.

Please note that the perf is not that bad - there is definitely no TPU->CPU->TPU roundtrips intended for prod - it is currently done as CPU->TPU via JAX for tests, while the long term goal is CPU->TPU via torch_tpu for tests, and no data transfers at all when the matmul consumes data that's already in the TPU's HBM, since in that case neither JAX nor torch_tpu will be used to transfer the data; instead, teh data will be from the :previous step". 

This is already a 500-line change, holding will cause maintance pain such as conatantly have to rebase, resolve merge conflicts, etc.

Please note that the actual code change will be minimal for the next step - only `jax.device_put/get` will be removed, and the device will be "xla"/"tpu" rather than "cpu". The vast majority of work will be to make sure that torch_tpu integrates with inductor and set up the env properly. Given this, and all above reasons, and that the TPU feature currently is not even meant to be used (thus the "not user facing" label), I'd prefer checking this in, have the accuracy baseline, and then swap in torch_tpu as a separate effort, instead of holding this giant PR and do tons of maintenance, plus mixed code changes that we don't even know if the accuracy baseline is quietly moved.

Thoughts?

Comment (jansel) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: At the very least this should also support GPU like the other code we have been adding does.  IMO correctness testing for non-templates is more valuable, since that exercise the actual codegen in inductor (as opposed to if you implemented the template correctly).

Comment (yarongmu-google) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: Thank you for this incredibly detailed and valuable feedback.  To summarize the above discussions, there are several "tracks" of works that need happen:

1. The "CPU-Tensor Hack" (Immediate Priority)
Like we discussed, I will switch to torch_tpu after this PR merges (mainly to avoid having to the overhead of constantly pull&merge etc etc). This is a env/set up change, with minimal code change once I verify the installation is successful and tensors can be directly created in the TPU's HBM. I anticipate the main difficulty is to find a stable (or a working) version to torch_tpu.

2. The "Generation Track" vs. "Competition Track" (Follow-up PR)
This PR focused on the "competition" track, namely followed the established paths of CUDA/Triton etc to register the Pallas TPU blocked matmul kernel to "compete" against the other kernels in the benchmarking arena. This work is orthogonal - as you rightly pointed out - to the generation of the kernel directly from aten ops. I chose to impl the competition track first, as it is "closer" to inductor (while "generation" is "closer" to helion). 

My point is that I completely agree with you, but generation should be its own dedicated, and quite diff PR. I plan to address this right after the integration of torch_tpu, at which point like you said we should create `pl.kernel` + `plgpu.as_torch_kernel`. I don't think - unless I missed something - what this PR covers will be wasted or dramatically changed - after all, hardcoded (in this PR) or generated (the future PR) Pallas kernels have to go through the same competition track pipes.

3. Unifying GPU/TPU Support (Future Work)
I would like to do to this after the generation work is done - Pallas has a completely different API set for GPUs; so we will need both the GPU generation path and TPU generation path in, before we "unify" them. That's why I left the GPU part unaddressed in this PR.

Does the above make sense?

In short, I agree with you, but I would like to tackle these step by step using diff PRs, each addressing only one item, instead of mingle them together into a giant PR, which will be hard to review, and even worse if somehow we will need rollbacks (we probably don't have this issue right now since TPU CI is not up yet but that's still a good habit to have). However, I merely chose a different implementation sequence (competition -> generation) vs what you had in mind (generation -> competition). 

I'd suggest let's get the competition in, so we can move on to (1) include torch_tpu and (2) add the generation path, like you rightly pointed out, instead of arguing about the sequence of implementations. Thoughts?

Comment (jansel) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: IMO supporting GPU would improve this PR and not be much code, since it would force the CPU-Tensor hack to be refactored into a common place (rather and copy-and-pasted into the template).

I guess I don't see the point of generating Pallas kernels that do CPU=>TPU=>CPU H2D copies on every kernel call.  It is just creating code we will need to delete later.

Even as a short term enablement thing, I don't see the point.  We can do the same enablement work on GPUs and have the tensors be on the right device.  GPUs let us test correctness better and more easily than using the wrong device.  IMO we should do TPUs right from the start.

Comment (yarongmu-google) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: Like I said above - competition and generation are 2 DIFFERENT workflows; both are equally valuable - why abandoning all this and switching to generation first, and then add back competition? Is it the plan that Pallas on TPU will not go through any competitions?

As for GPU - Pallas has a completely different set of APIs for it, and IIUC Oguz has been adding GPU support. In that case, I can hold this until his GPU changes for matmul are in?

Comment (jansel) [reply torch/_inductor/codegen/pallas_kernels/block_mm_template.py]: Yes, I am aware they are different -- but part of what PyTorch does is unify many different lower level API to make programming different devices similar.  We need to build abstractions to support both, not just create duplicate code.
